### PR TITLE
fix(backend): recognize "sidecar.istio.io/inject" annotation to use Istio service mesh optionally. Fixes #6200

### LINKED ISSUE #6067: Istio injection always disabled
While trying to put together a Kubeflow distribution that includes as many as possible elements inside the Istio mesh, we found that for Kubeflow Pipelines steps the istio sidecar injection is always overriden to "false". 

(This would for example make it impossible for a pipelines step to make calls to a KFServing model in the same namespace which has an istio sidecar enabled).

This seems to be hardcoded here:

https://github.com/kubeflow/pipelines/blob/fff62e1521e89dd31c87eec03073b8f8627d73a3/backend/src/apiserver/resource/resource_manager.go#L384

I presume that was done because the Pipelines owner are very sure that KFP would not work with injection enabled. As far as I know, newer versions of Argo Workflows should now support it.

Could we turn this into an ENV var?

### PR DESCRIPTION
**Pull request by Matts966**
**Description of your changes:**

Fixes #6067, #6200

Instead of always overwriting `"sidecar.istio.io/inject"` annotation in runtime, set the annotation if the key does not exist.

By this change, users can use Istio service mesh optionally to enable communication between pipelines and another micro services. Without this change, the communication can cause errors like `RBAC: access denied` even with appropriate routing.

In pipeline SDK, enabling Istio service mesh is as easy as

```python
@kfp.dsl.pipeline(name=PIPELINE_NAME)
def dpf_pipeline():
    debug_task = debug_op()
    # Inject Istio proxy
    debug_task.add_pod_annotation("sidecar.istio.io/inject", "true")
    # Wait Istio proxy setup
    debug_task.add_pod_annotation("proxy.istio.io/config", '{ "holdApplicationUntilProxyStarts": true }')
```

As written in [doc on Argo Repository](https://github.com/argoproj/argo-workflows/blob/master/docs/sidecar-injection.md), Istio sidecar injection can be problematic for controlling entire workflow, however, users can optionally set container lifecycle like below to terminate Istio sidecar proxy.

```yaml
spec:
  containers:
  - name: lifecycle-example
    image: pipeline-conatiner-image
    lifecycle:
      preStop:
        exec:
          command: ["/bin/sh","-c","curl -XPOST localhost:15000/quitquitquit"]
```

In addition, in our product, we do not experience such an endless sidecar containers. Istio proxies exits by themselves when other containers finish and will be `COMPLETED` state.

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Matts966** [comment]
- Always injecting Istio sidecar can be problematic
  - For performance reason. All pipelines should wait Istio proxy setup for safety reason, because without the proxy, all network actions will be buggy. 
  - Also, as the reason above, we should additionally annotate `"proxy.istio.io/config"`. This leads to more complexity.

- Without this CL, current Kubeflow SDK's behavior is mysterious for users because the sample code above will be overwritten mysteriously in runtime.
- `SetAnnotationsToAllTemplates` function is called only for setting `"sidecar.istio.io/inject"`. So modifying is relatively safe.

File: backend/src/common/util/workflow.go [hunk 1]
```diff
@@ -202,15 +202,19 @@ func (w *Workflow) OverrideName(name string) {
...
+func (w *Workflow) SetAnnotationsToAllTemplatesIfKeyNotExist(key string, value string) {
 	if len(w.Spec.Templates) == 0 {
 		return
 	}
 	for index := range w.Spec.Templates {
 		if w.Spec.Templates[index].Metadata.Annotations == nil {
 			w.Spec.Templates[index].Metadata.Annotations = make(map[string]string)
 		}
+		if _, isSet := w.Spec.Templates[index].Metadata.Annotations[key]; isSet {
+			return
```
Comment (Bobgy) [inline backend/src/common/util/workflow.go:216]: nit: I believe you want to write `continue` here, is this a mistake?

**Review by Bobgy — COMMENTED** [review]
/lgtm
/ok-to-test
Thank you for the contribution! Yes, I fully agree with your rationale.

**Comment by Bobgy** [comment]
Supercedes https://github.com/kubeflow/pipelines/pull/6200

File: backend/src/common/util/workflow.go (hunk 1 above)
Comment (Matts966) [reply backend/src/common/util/workflow.go]: Yes, you are right. This made our pipeline's steps always use Istio and broke some ops using Workload Identity yesterday. I fixed locally and forgot. Sorry for my broken patch.

I confirmed this patch has no diff from our latest env now.

Comment (Matts966) [reply backend/src/common/util/workflow.go]: BTW, the failure of Workload Identity was fixed by annotating `"traffic.sidecar.istio.io/excludeOutboundIPRanges=169.254.169.254/32"`.

https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#fails

@Bobgy 🙇‍♂️

**Pull request merged**